### PR TITLE
Fix: Auth skip

### PR DESCRIPTION
**Pull request by Meldiron**
- Better reversion mechanism for skip() method
- Improved docs
- Improved error message

Let's imagine the following pseudo code:
```php
Auth::disable();

$doc = $db->getDocument('users', 'user1'); // Works

$doc = Auth::skip(fn() => $db->getDocument('users', 'user1')); // Works

$doc = $db->getDocument('users', 'user1'); // No longer work
```

If you disable auth, and later use `skip`, it re-enabled auth. After `skip` is finished, the status should be restored to what it was before skipping, instead of enabling it forcefully. 
Can be seen on: https://github.com/utopia-php/database/pull/130/files#diff-f0b0bb401b111034042efdb8f446197096a3fa9592d55d36ca9c6213770e5e25L161

### CODE REVIEW

File: src/Database/Validator/Authorization.php [hunk 1]
```diff
@@ -8,24 +8,31 @@
 class Authorization extends Validator
 {
     /**
+     * Array of allowed roles.
+     * 
      * @var array
      */
     static $roles = ['role:all' => true];
```
Comment (TorstenDittmann) [inline src/Database/Validator/Authorization.php:15]:
```suggestion
    static array $roles = ['role:all' => true];
```

File: src/Database/Validator/Authorization.php [hunk 2]
```diff
@@ -8,24 +8,31 @@
...
 
     /**
+     * Action of the authorization validation.
+     * Mainly 'read' or 'write'.
+     * 
      * @var string
      */
     protected $action = '';
```
Comment (TorstenDittmann) [inline src/Database/Validator/Authorization.php:23]:
```suggestion
    protected string $action = '';
```

File: src/Database/Validator/Authorization.php [hunk 3]
```diff
@@ -148,23 +173,25 @@ public static function setDefaultStatus(bool $status): void
...
     /**
-     * Skip Authorization
+     * Skip Authorization for the code executed inside the callback.
      * 
-     * Skips authorization for the code to be executed inside the callback
+     * @param callable $callback function to run without authorization
      * 
      * @return mixed
      */
     public static function skip(callable $callback)
```
Comment (TorstenDittmann) [inline src/Database/Validator/Authorization.php:182]:
```suggestion
    public static function skip(callable $callback): mixed
```

**Review by TorstenDittmann — COMMENTED** [review]
Try to also fill out these types when missing 👍🏻

**Comment by Meldiron** [comment]
Oooh now I see what you mean. All updated.

**Comment by eldadfux** [comment]
@Meldiron it would have been very helpful to get some background and context in the PR description. What has caused us to change this mechanism?

Since this is a very sensitive security method in the entire architecture any extra information will help review the changes with more confidence.

**Comment by Meldiron** [comment]
@eldadfux Updated. What we did until now was a bug I believe.

**Comment by eldadfux** [comment]
@Meldiron please carefully test this doesn't break anything in Appwrite before we can merge it.

**Comment by eldadfux** [comment]
Any chance this was already solved at #130 ?

**Comment by eldadfux** [comment]
After checking with @abnegate, this was fixed and merged here: https://github.com/utopia-php/database/pull/186

**Pull request closed without merging**